language-plutus-core/src/Language/PlutusCore.hs

```diff
@@ -160,24 +163,36 @@ checkFile :: FilePath -> IO (Maybe T.Text)
 checkFile = fmap (either (pure . prettyText) id . fmap (fmap prettyPlcDefText . check) . parse) . BSL.readFile
 
 -- | Print the type of a program contained in a 'ByteString'
-printType :: (AsParseError e AlexPosn, AsRenameError e AlexPosn, AsTypeError e AlexPosn, MonadError e m) => BSL.ByteString -> m T.Text
+printType
+    :: (AsParseError e AlexPosn, AsUniqueError e AlexPosn, AsRenameError e AlexPosn, AsTypeError e AlexPosn, MonadError e m)
+    => BSL.ByteString
+    -> m T.Text
 printType = printNormalizeType False
 
 -- | Print the type of a program contained in a 'ByteString'
-printNormalizeType :: (AsParseError e AlexPosn, AsRenameError e AlexPosn, AsTypeError e AlexPosn, MonadError e m) => Bool -> BSL.ByteString -> m T.Text
+printNormalizeType
+    :: (AsParseError e AlexPosn, AsUniqueError e AlexPosn, AsRenameError e AlexPosn, AsTypeError e AlexPosn, MonadError e m)
+    => Bool
+    -> BSL.ByteString
+    -> m T.Text
 printNormalizeType norm bs = runQuoteT $ prettyPlcDefText <$> do
     scoped <- parseScoped bs
     annotated <- annotateProgram scoped
     typecheckProgram (TypeCheckCfg 1000 $ TypeConfig norm mempty) annotated
 
 -- | Parse and rewrite so that names are globally unique, not just unique within
 -- their scope.
-parseScoped :: (AsParseError e AlexPosn, MonadError e m) => BSL.ByteString -> m (Program TyName Name AlexPosn)
-parseScoped str = runQuoteT $ parseProgram str >>= rename
+parseScoped
+    :: (AsParseError e AlexPosn, AsUniqueError e AlexPosn, MonadError e m)
+    => BSL.ByteString
+    -> m (Program TyName Name AlexPosn)
+-- don't require there to be no free variables at this point, we might be parsing an open term
+parseScoped = runQuoteT . (through (Uniques.checkProgram (\case { FreeVariable {} -> False; _ -> True })) <=< rename <=< parseProgram)
```

Comment:
Can we have two versions - one which does sanity checking and one which doesn't?

Reply:
We can do, although I don't see much reason to turn it off at this stage!